django/db/models/query_utils.py

```diff
@@ -203,7 +204,22 @@ def __get__(self, instance, cls=None):
                     raise AttributeError(
                         "Cannot read a generated field from an unsaved model."
                     )
-                instance.refresh_from_db(fields=[field_name])
+
+                def fetch_for_instances(instances):
+                    if len(instances) == 1:
+                        instances[0].refresh_from_db(fields=[field_name])
+                    else:
+                        attname = self.field.attname
+                        value_by_pk = {
+                            pk: value
+                            for pk, value in self.field.model._base_manager.db_manager()
+                            .filter(pk__in={i.pk for i in instances})
+                            .values_list("pk", attname)
+                        }
+                        for instance in instances:
+                            setattr(instance, attname, value_by_pk[instance.pk])
+
+                get_lazy_mode()(instance, self.field, fetch_for_instances)
```

Comment:
Yeah, that would be good. I originally tried an interface but it was hard for some reason. At least, it was confusing whether methods should live on the descriptors or fields. I will give it another try, making the descriptors the “fetchers”.

Fetch one and many have separate implementations here only for backwards compatibility. Since its introduction, `refresh_from_db()` [has been documented](https://docs.djangoproject.com/en/5.0/ref/models/instances/#django.db.models.Model.refresh_from_db) as the backing method for deferred attributes:

> Note that when deferred fields are accessed, the loading of the deferred field’s value happens through this method.

There are more backwards compatibility concerns for related descriptors, but there won’t be for `GenericForeignKey` (TODO), so maybe we should only have one method.

Reply:
That looks great! Any reason for going with a single `fetch` method over two distinct `fetch_(one|many)`?

I don't see it as a hard blocker but the fact that all `fetch` implementations have an `if` branch that is unnecessary from the caller's perspective (the mode handler knows which method it should call) seemed odd to me.